src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs

```diff
@@ -258,5 +260,192 @@ public void BackoffSupervisor_must_support_manual_reset()
+        }
+
+        [Fact]
+        public void BackoffSupervisor_not_reply_to_sender_if_replyWhileStopped_is_not_specified()
+        {
+            EventFilter.Exception<TestException>().Expect(1, () =>
+            {
+                var supervisor = Create(Backoff.OnFailure(Child.Props(TestActor), "c1", TimeSpan.FromSeconds(100), TimeSpan.FromSeconds(300), 0.2, -1));
+                supervisor.Tell(BackoffSupervisor.GetCurrentChild.Instance);
+
+                var c1 = ExpectMsg<BackoffSupervisor.CurrentChild>().Ref;
+                Watch(c1);
+                supervisor.Tell(BackoffSupervisor.GetRestartCount.Instance);
+                ExpectMsg<BackoffSupervisor.RestartCount>().Count.Should().Be(0);
+
+                c1.Tell("boom");
+                ExpectTerminated(c1);
+
+                AwaitAssert(() =>
+                {
+                    supervisor.Tell(BackoffSupervisor.GetRestartCount.Instance);
+                    ExpectMsg<BackoffSupervisor.RestartCount>().Count.Should().Be(1);
+                });
+
+                supervisor.Tell("boom"); //this will be sent to deadLetters
+                ExpectNoMsg(500);
+            });
+        }
+
+        [Theory, ClassData(typeof(DelayTable))]
+        public void BackoffSupervisor_correctly_calculate_the_delay(int restartCount, TimeSpan minBackoff, TimeSpan maxBackoff, double randomFactor, TimeSpan expectedResult)
+        {
+            Assert.Equal(expectedResult, BackoffSupervisor.CalculateDelay(restartCount, minBackoff, maxBackoff, randomFactor));
+        }
+
+        internal class DelayTable : IEnumerable<object[]>
+        {
+            private readonly List<object[]> delayTable = new List<object[]>
+            {
+                new object[] { 0, TimeSpan.Zero, TimeSpan.Zero, 0.0, TimeSpan.Zero },
+                new object[] { 0, TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(7), 0d, TimeSpan.FromMinutes(5) },
+                new object[] { 2, TimeSpan.FromSeconds(5), TimeSpan.FromSeconds(7), 0d, TimeSpan.FromSeconds(7) },
+                new object[] { 2, TimeSpan.FromSeconds(5), TimeSpan.FromDays(7), 0d, TimeSpan.FromSeconds(20) },
+                new object[] { 29, TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(10), 0d, TimeSpan.FromMinutes(10) },
+                new object[] { 29, TimeSpan.FromDays(10000), TimeSpan.FromDays(10000), 0d, TimeSpan.FromDays(10000) },
+                new object[] { int.MaxValue, TimeSpan.FromDays(10000), TimeSpan.FromDays(10000), 0d, TimeSpan.FromDays(10000) }
+            };
+
+            public IEnumerator<object[]> GetEnumerator() => delayTable.GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        [Fact]
+        public void BackoffSupervisor_stop_restarting_the_child_after_reaching_maxNrOfRetries_limit_using_BackOff_OnStop()
+        {
+            var supervisor = Create(OnStopOptions(maxNrOfRetries: 2));
+
+            IActorRef WaitForChild(int n = 0)
```

Comment:
Do we have anything like `scalatest.concurrent.Eventually`? That would simplify this piece of code.

Reply:
`AwaitCondition` ?

Reply:
Exactly that.